ddht/tools/lru_sql_dict.py

```diff
@@ -0,0 +1,271 @@
+
+        else:
+            # evict least recently used key/value
+            if self.is_full:
+                self.__delitem__(self.tail.key)
+
+            # get old head
+            old_head_pkey, _, _, old_head_pref, _ = self.head
+
+            # get attributes for new head
+            (max_pkey,) = self._fetch_single_query(
+                "SELECT id FROM cache ORDER BY id DESC LIMIT 1;"
+            )
+            new_pkey = max_pkey + 1
+            new_pref = None
+            new_nref = old_head_pkey
+
+            # add new head to cache
+            self._execute(
+                "INSERT INTO cache VALUES (?, ?, ?, ?, ?);",
+                (new_pkey, key, pickled_value, new_pref, new_nref),
+            )
+
+            # add new head to map
+            self._execute(
+                "INSERT INTO map VALUES (?, ?);", (key, new_pkey),
+            )
+
+            # update old head in cache
+            self._execute(
+                "UPDATE cache SET pref=? WHERE id=?;", (new_pkey, old_head_pkey),
+            )
+
+    def _update_item(self, key: Union[str, bytes], value: Any) -> None:
+        pickled_value = self._pickle(value)
+        self._execute("UPDATE cache SET value=? WHERE key=?;", (pickled_value, key))
+
+    def _fetch_single_query(self, query: str, args: Tuple[Any, ...] = None) -> Any:
+        with self._conn:
+            if args:
+                cursor = self._conn.execute(query, args).fetchall()
+            else:
+                cursor = self._conn.execute(query).fetchall()
+        if len(cursor) > 1:
+            raise Exception(
+                f"Invalid db state. More than one result found for query: {query}."
+            )
+        if not cursor:
+            return None
+        return first(cursor)
+
+    def _execute(self, query: str, args: Tuple[Any, ...] = None) -> None:
+        with self._conn:
+            if args:
+                self._conn.execute(query, args)
+            else:
+                self._conn.execute(query)
+
+    def _pickle(self, value: Any) -> bytes:
+        return sqlite3.Binary(pickle.dumps(value, -1))
+
+    def _unpickle(self, value: bytes) -> Any:
+        return pickle.loads(value)
+
+    @property
+    def is_full(self) -> bool:
+        return self.__len__() >= self.cache_size
+
+    @property
+    def is_empty(self) -> bool:
+        return self.__len__() == 0
+
+    @property
+    def head(self) -> Node:
+        head = self._fetch_single_query(
+            "SELECT id,key,value,pref,nref FROM cache WHERE pref IS NULL;"
+        )
+        if not head:
+            raise Exception("No head found.")
+        return Node(*head)
+
+    @property
+    def tail(self) -> Node:
+        tail = self._fetch_single_query(
+            "SELECT id,key,value,pref,nref FROM cache WHERE nref IS NULL;"
+        )
+        if not tail:
+            raise Exception("No tail found.")
+        return Node(*tail)
+
+    # custom iterator type not compatible with supertype
```

Comment:
This seems like it should be something you can satisfy.  You probably need to update this class to use [generic typing](https://docs.python.org/3/library/typing.html#user-defined-generic-types)

Reply:
I'm having a little trouble with this one. I've updated the types (for `items()` and `values()`) to be compatible with the supertype's signature, which is defined with `ValueViews` and `AbstractSet` - but then mypy complains that it is expecting a `Generator` type due to the `yield`. I haven't been able to find some way to reconcile these different types, and it's not obvious to me how I could achieve this using generics.

Reply:
I'm not sure it is worth doing but:

```python
In [41]: collections.abc.ValuesView({'a': 1, 'b': 2})
Out[41]: ValuesView({'a': 1, 'b': 2})

In [42]: tuple(_41)
Out[42]: (1, 2)
```

It looks like you can instantiate your own `ValuesView` but it also looks like it'd be quite awkward to make this work with that API....  So :+1: on just `# type: ignore` but it would be cool if there was a clean way to do this.